pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,709 @@
+
+		/// Yield per period can't be 0
+		InvalidYieldPerPeriod,
+
+		/// Provided total_rewards for farming is bellow min limit
+		InvalidTotalRewards,
+
+		/// Reward currency balance too low
+		InsufficientRewardCurrencyBalance,
+
+		/// Account is not allowed to perform action
+		Forbidden,
+
+		/// Farm does not exist
+		FarmNotFound,
+
+		/// Liquidity pool already exist in farm
+		LiquidityPoolAlreadyExists,
+
+		/// Weight can't be 0
+		InvalidWeight,
+
+		/// Loyalty curver b param should be from [0, 1)
+		InvalidLoyaltyCurverParamB,
+
+		/// AMM pool does not exist
+		AmmPoolDoesNotExist,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// New farm was creaated by `CreateOrigin` origin. [pool_id, global_pool]
+		FarmCreated(
+			PoolId,
+			GlobalPool<T::BlockNumber, Balance, T::CurrencyId, T::AccountId, T::BlockNumber>,
+		),
+
+		/// New liquidity(AMM) pool was added to farm [farm_id, amm_pool_id, liquidity_pool]
+		LiquidityPoolAdded(PoolId, T::AccountId, LiquidityPool<T::BlockNumber, Balance>),
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn pool_id)]
+	pub type PoolIdSeq<T: Config> = StorageValue<_, PoolId, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn global_pool)]
+	pub type GlobalPoolData<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		PoolId,
+		GlobalPool<T::BlockNumber, Balance, T::CurrencyId, T::AccountId, T::BlockNumber>,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn liquidity_pool)]
+	pub type LiquidityPoolData<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		PoolId,
+		Twox64Concat,
+		T::AccountId, //amm_pool_id
+		LiquidityPool<T::BlockNumber, Balance>,
+		OptionQuery,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create new liquidity mining program
+		///
+		/// Parameters:
+		/// - `origin`:
+		/// - `total_rewards`:
+		/// - `planned_yielding_periods`: planned number of periods to distribute rewards. WARN: this is not
+		/// how long will farming run.  Owner can destroy farm sooner or liq. mining can run longer
+		/// if all the rewards will not distributed.
+		/// - `blocks_per_period`:
+		/// - `incetivized_token`
+		/// - `reward_currency`
+		/// - `admin_account`
+		/// - `yield_per_period`
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn create_farm(
+			origin: OriginFor<T>,
+			total_rewards: Balance,
+			planned_yielding_periods: T::BlockNumber,
+			blocks_per_period: T::BlockNumber,
+			incentivized_token: T::CurrencyId,
+			reward_currency: T::CurrencyId,
+			owner: T::AccountId,
+			yield_per_period: Permill,
+		) -> DispatchResult {
+			T::CreateOrigin::ensure_origin(origin)?;
+
+			Self::validate_create_farm_data(
+				total_rewards,
+				planned_yielding_periods,
+				blocks_per_period,
+				yield_per_period,
+			)?;
+
+			ensure!(
+				T::MultiCurrency::free_balance(reward_currency, &owner) >= total_rewards,
+				Error::<T>::InsufficientRewardCurrencyBalance
+			);
```

Comment:
I'd do this before validating the farm data.

Reply:
I thought that `free_balance()` is doing storage read, that's why I used this order.